addons/controllers/clusterbootstrap_controller.go

```diff
@@ -300,79 +343,371 @@ func (r *ClusterBootstrapReconciler) createOrPatchclusterBootstrapFromTemplate(
+		pkgi.Spec.PackageRef = &kapppkgiv1alpha1.PackageRef{
+			// clusterBootstrap.Spec.Kapp.RefName is Package.Name. I.e., kapp-controller.tanzu.vmware.com.0.28.0+vmware.1-tkg.1-rc.1
+			// PackageInstall.Spec.PackageRef looks for the Package.Spec.refName which is a short name of the full Package.Name
+			// packageRefName and packageVersion are fetched from the Package CR.
+			RefName: packageRefName,
+			VersionSelection: &versions.VersionSelectionSemver{
+				Constraints: packageVersion,
+				Prereleases: &versions.VersionSelectionSemverPrereleases{},
+			},
+		}
+		// Adding the cluster reference to PackageInstall spec to instruct kapp-controller where to deploy
+		// the underlying resources
+		clusterKubeconfigDetails := util.GetClusterKubeconfigSecretDetails(cluster)
+		pkgi.Spec.Cluster = &kappctrlv1alpha1.AppCluster{
+			KubeconfigSecretRef: &kappctrlv1alpha1.AppClusterKubeconfigSecretRef{
+				Name: clusterKubeconfigDetails.Name,
+				Key:  clusterKubeconfigDetails.Key,
+			},
+		}
+		secretName, err := r.GetDataValueSecretNameFromBootstrapPackage(clusterBootstrap.Spec.Kapp, cluster.Namespace)
 		if err != nil {
-			log.Error(err, "failed to getGVR")
-			return nil, err
+			return err
+		}
+		pkgi.Spec.Values = []kapppkgiv1alpha1.PackageInstallValues{
+			{SecretRef: &kapppkgiv1alpha1.PackageInstallValuesSecretRef{
+				Name: secretName},
+			},
 		}
-		provider, err := r.dynamicClient.Resource(*gvr).Namespace(namespace).Get(r.context, pkg.ValuesFrom.ProviderRef.Name, metav1.GetOptions{}, "status")
+		return nil
+	}
 
-		if err != nil {
-			log.Error(err, "unable to fetch provider", "provider", pkg.ValuesFrom.ProviderRef, "gvr", gvr)
-			return nil, err
+	_, err = controllerutil.CreateOrPatch(r.context, r.Client, pkgi, pkgiMutateFn)
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("unable to create or patch %s PackageInstall CR for cluster: %s",
+			constants.KappControllerAddonName, cluster.Name))
+		return err
+	}
+
+	return nil
+}
+
+// createOrPatchPackageOnRemote creates the Package CR on remote cluster. In order to install a package on remote cluster
+// the Package CR needs to be present.
+// createOrPatchPackageOnRemote returns a tuple: (<remote-package>, <error>)
+func (r *ClusterBootstrapReconciler) createOrPatchPackageOnRemote(cluster *clusterapiv1beta1.Cluster,
+	cbPkg *runtanzuv1alpha3.ClusterBootstrapPackage, clusterClient client.Client) (*kapppkgv1alpha1.Package, error) {
+
+	var err error
+	// Create or patch Package CR on remote cluster
+	localPackage := &kapppkgv1alpha1.Package{}
+	key := client.ObjectKey{Namespace: cluster.Namespace, Name: cbPkg.RefName}
+	if err = r.Client.Get(r.context, key, localPackage); err != nil {
+		// If there is an error to get the Carvel Package CR from local cluster, nothing needs to be created/cloned on remote.
+		// Let the reconciler try again.
+		r.Log.Error(err, fmt.Sprintf("unable to get package %s, nothing needs to be created on cluster %s/%s",
```

Comment:
Somehow the part for ```nothing needs to be created on cluster``` in the error message seems misleading to me

Reply:
Re-phrased it a little bit: `unable to create or patch Package %s on cluster %s/%s. Error occurs when getting Package %s from the management cluster`

Reply:
+1, err is not special here